bin/liquidator/src/tx_sender.rs

```diff
@@ -0,0 +1,237 @@
+    tx_liq_trigger_receiver: Receiver<()>,
+    tx_tcs_trigger_receiver: Receiver<()>,
+    tx_tcs_trigger_sender: Sender<()>,
+    rebalance_trigger_sender: Sender<()>,
+    shared_state: Arc<RwLock<SharedState>>,
+    liquidation: Box<LiquidationState>,
+    tcs: Box<TcsState>,
+) -> Vec<JoinHandle<()>> {
+    if max_parallel_operations < 1 {
+        error!("max_parallel_operations must be >= 1");
+        std::process::exit(1)
+    }
+
+    let reserve_one_worker_for_liquidation = max_parallel_operations > 1 && enable_liquidation;
+    let tcs_capable_worker_count = (max_parallel_operations as usize)
+        - if reserve_one_worker_for_liquidation {
+            1
+        } else {
+            0
+        };
+
+    let workers: Vec<JoinHandle<()>> = (0..max_parallel_operations)
+        .map(|worker_id| {
+            tokio::spawn({
+                let shared_state = shared_state.clone();
+                let receiver_liq = tx_liq_trigger_receiver.clone();
+                let receiver_tcs = tx_tcs_trigger_receiver.clone();
+                let sender_tcs = tx_tcs_trigger_sender.clone();
+                let rebalance_trigger_sender = rebalance_trigger_sender.clone();
+                let liquidation = liquidation.clone();
+                let tcs = tcs.clone();
+                async move {
+                    worker_loop(
+                        shared_state,
+                        receiver_liq,
+                        receiver_tcs,
+                        sender_tcs,
+                        rebalance_trigger_sender,
+                        liquidation,
+                        tcs,
+                        worker_id,
+                        tcs_capable_worker_count,
+                        reserve_one_worker_for_liquidation && worker_id == 0,
+                    )
+                    .await;
+                }
+            })
+        })
+        .collect();
+
+    workers
+}
+
+async fn worker_loop(
+    shared_state: Arc<RwLock<SharedState>>,
+    liq_receiver: Receiver<()>,
+    tcs_receiver: Receiver<()>,
+    tcs_sender: Sender<()>,
+    rebalance_trigger_sender: Sender<()>,
+    mut liquidation: Box<LiquidationState>,
+    mut tcs: Box<TcsState>,
+    id: u64,
+    tcs_capable_workers: usize,
+    only_liquidation: bool,
+) {
+    loop {
+        debug!(
+            "Worker #{} waiting for task (only_liq={})",
+            id, only_liquidation
+        );
+
+        let _ = if only_liquidation {
+            liq_receiver.recv().await.expect("receive failed")
+        } else {
+            tokio::select!(
+                _ = liq_receiver.recv() => {},
+                _ = tcs_receiver.recv() => {},
+            )
+        };
+
+        // a task must be available to process
+        // find it in global shared state, and mark it as processing
+        let task = worker_pull_task(&shared_state, id, tcs_capable_workers, only_liquidation)
+            .expect("Worker woke up but has nothing to do");
```

Comment:
I think with the tcs batching pulled back in, this no longer works. You send a trigger for each tcs candidate, but the processing tasks eat up multiple at once - so for N candidates the there'll be fewer than N worker tasks.

There's also some weirdness now where the tcs jobs trigger on the tcs channel and take all the liquidation jobs and then the liquidation job wakes up (on the liq channel) and has no work to do.

It may be more robust to avoid the requirement that the number of messages on the channels and the number of work chunks match up exactly.

Reply:
It "works" because the TCS task will just be made of an empty vector (no candidate) if everything was consumed before

Reply:
Ah, I see! Would you be up for making it more explicit by adding a no-work WorkerTask?